ipq806x: add support for OpenMesh A42

ecsv opened the pull request:
OpenMesh A42 is a dual-band () router, based on QCA IPQ401x. More information can be found at https://wikidevi.com/wiki/Open_Mesh_A42

* QCA IPQ401x
* 256 MB of RAM
* 32 MB of SPI NOR flash (s25fl256s1)
  - 2x 15 MB available; but one of the 15 MB regions is the recovery image
* 2T2R 2.4 GHz
  - QCA4019 hw1.0 (SoC)
  - requires special BDF in QCA4019/hw1.0/board-2.bin with
    bus=ahb,bmi-chip-id=0,bmi-board-id=16,variant=OM-A42
* 2T2R 5 GHz
  - QCA4019 hw1.0 (SoC)
  - requires special BDF in QCA4019/hw1.0/board-2.bin with
    bus=ahb,bmi-chip-id=0,bmi-board-id=17,variant=OM-A42
* multi-color LED (controlled via red/green/blue GPIOs)
* 1x button (reset; kmod-input-gpio-keys compatible)
* external watchdog
  - triggered GPIO
* 1x USB (xHCI)
* TTL pins are on board (arrow points to VCC, then follows: GND, TX, RX)
* 2x gigabit ethernet
* powered only via POE
  - 802.3af POE on Ethernet 1
  - 18-24v passive POE (mode B) on Ethernet 2

The tool ap51-flash (https://github.com/ap51-flash/ap51-flash) should be
used to transfer the factory image to the u-boot when the device boots up.

---

The u-boot is currently only available from https://drive.google.com/drive/folders/1lagQhi80R3UMnY_8qo_5zC02gbLpSpYZ?usp=sharing

The upstreaming progress of the ath10k "qcom,ath10k-calibration-variant" patch can be found in https://patchwork.kernel.org/patch/10102249/ (it was applied)

The upstreaming progress of the BDFs can be found in http://lists.infradead.org/pipermail/ath10k/2017-November/010470.html (state unknown)

----

This PR can originally be found at https://github.com/lede-project/source/pull/1550

@mkresin provided some feedback. I have integrated them and the end result has following things changed:

* ordered items in cases alphabetically (broke when switching from a42 to openmesh,a42)
  * actually, mkresin did that
* removed uci entries for unused LEDs
  * actually, mkresin did that
* switched sysupgrade from combined extended images to sysupgrade.tar
  * actually, mkresin did the main work and I've just fixed some bugs
* switched from om-watchdog to gpio-wdt
  * Thanks mkresin for the hint
* switched from gzip compressed kernel to LZMA
* switched TARGET_DEVICE name to openmesh_a42
  * actually, mkresin did that
* don't include qcom-ipq4019-ap.dk01.1.dtsi anymore in DTS
  * required a lot of copying from this file to the a42 dts
* padded kernel to erase boundary to make sure that 0xDEADC0DE EOF marker is correctly aligned for rootfs(_data)
  * otherwise the JFFS2 will not be correctly created after sysupgrade
* merged all ipq806x patches for A42 in a single patch
  * actually, mkresin did that

ecsv committed:
uboot-envtools: add OpenMesh A42 support

Signed-off-by: Sven Eckelmann <sven.eckelmann@open-mesh.com>

ecsv committed:
ipq806x: add support for OpenMesh A42

* QCA IPQ401x
* 256 MB of RAM
* 32 MB of SPI NOR flash (s25fl256s1)
  - 2x 15 MB available; but one of the 15 MB regions is the recovery image
* 2T2R 2.4 GHz
  - QCA4019 hw1.0 (SoC)
  - requires special BDF in QCA4019/hw1.0/board-2.bin with
    bus=ahb,bmi-chip-id=0,bmi-board-id=16,variant=OM-A42
* 2T2R 5 GHz
  - QCA4019 hw1.0 (SoC)
  - requires special BDF in QCA4019/hw1.0/board-2.bin with
    bus=ahb,bmi-chip-id=0,bmi-board-id=17,variant=OM-A42
* multi-color LED (controlled via red/green/blue GPIOs)
* 1x button (reset; kmod-input-gpio-keys compatible)
* external watchdog
  - triggered GPIO
* 1x USB (xHCI)
* TTL pins are on board (arrow points to VCC, then follows: GND, TX, RX)
* 2x gigabit ethernet
* powered only via POE
  - 802.3af POE on Ethernet 1
  - 18-24v passive POE (mode B) on Ethernet 2

The tool ap51-flash (https://github.com/ap51-flash/ap51-flash) should be
used to transfer the factory image to the u-boot when the device boots up.

Signed-off-by: Sven Eckelmann <sven.eckelmann@open-mesh.com>

ecsv committed:
build: add image command for CE images

Combined Extended Images V1 can be created easily via the new image
commands using

    IMAGE/sysupgrade.bin/squashfs := append-rootfs | pad-rootfs | combined-ext-image

Signed-off-by: Sven Eckelmann <sven.eckelmann@open-mesh.com>

ecsv committed:
ipq-wifi: add board-2.bin for OpenMesh A42

Signed-off-by: Sven Eckelmann <sven.eckelmann@open-mesh.com>

ecsv committed:
om-fwupgradecfg-gen.sh: add support for the A42

Signed-off-by: Sven Eckelmann <sven.eckelmann@open-mesh.com>

ecsv committed:
mac80211: ath10k: search DT for BDF variant info

Board Data File (BDF) is loaded upon driver boot-up procedure. The right
board data file is identified on QCA4019 using bus, bmi-chip-id and
bmi-board-id.

The problem, however, can occur when the (default) board data file cannot
fulfill the vendor requirements and it is necessary to use a different
board data file.

This problem was solved for SMBIOS by adding a special SMBIOS type 0xF8.
Something similar has to be provided for systems without SMBIOS but with
device trees. No solution was specified by QCA and therefore a new one has
to be found for ath10k.

The device tree requires addition strings to define the variant name

    wifi@a000000 {
    	status = "okay";
    	qcom,ath10k-calibration-variant = "RT-AC58U";
    };

    wifi@a800000 {
    	status = "okay";
    	qcom,ath10k-calibration-variant = "RT-AC58U";
    };

This would create the boarddata identifiers for the board-2.bin search

 *  bus=ahb,bmi-chip-id=0,bmi-board-id=16,variant=RT-AC58U
 *  bus=ahb,bmi-chip-id=0,bmi-board-id=17,variant=RT-AC58U

Signed-off-by: Sven Eckelmann <sven.eckelmann@openmesh.com>

ecsv committed:
combined-ext-image.sh: generate image in temp dir

The new build commands operate on the input image and use it again as
output image. This conflicts with the way combined-ext-image.sh was
operating. It required that input and output files are different files and
and that it can write freely to the output file.

This can be avoided when all intermediate build steps by
combined-ext-image.sh are done in a temporary directory. The output file is
then only overwritten in the last step.

Signed-off-by: Sven Eckelmann <sven.eckelmann@open-mesh.com>

ecsv committed:
ipq806x: enable gpio watchdog support

The kernel driver gpio-wdt or the userspace tool om-watchdog can be used to
trigger external gpio watchdog chips. The gpio-wdt driver has the benefit
that it can be configured together with the rest of the device in the DTS
and better integrates in the OpenWrt via procd.

Signed-off-by: Sven Eckelmann <sven.eckelmann@openmesh.com>

mkresin commented on include/image-commands.mk:
```diff
@@ -229,6 +229,19 @@ define Build/combined-image
 	@mv $@.new $@
 endef
 
+define Build/combined-ext-image
```
It is openmesh specific, isn't it? Please use a name which indicates it.

ecsv replied:
It was never planned to be OpenMesh specific but it looks like it ended up that way. What about openmesh-ce as name?

mkresin replied:
Or just openmesh-image, following the naming of already existing recipes in the file.

mkresin commented on include/image-commands.mk:
```diff
@@ -229,6 +229,19 @@ define Build/combined-image
 endef
 
+define Build/combined-ext-image
+	-sh $(TOPDIR)/scripts/om-fwupgradecfg-gen.sh \
```
Do not silence the return code. I guess we would like to know if something went wrong. Remove the minus!

Does it work as well if you don't specific the to be used shell? Just the script path and done.

mkresin commented on target/linux/ipq806x/base-files/lib/upgrade/openmesh.sh:
```diff
@@ -0,0 +1,87 @@
+
+	# switch to second kernel when flashing to secondary flash region
+	local next_boot_part="1"
+	case "$(board_name)" in
```
I have to admit, I'm still not sure if I get why you can't use the bootseq u-boot variable. Furthermore I don't really the hardcoded mtd index.

If I understand the mail you send to me correctly, the u-boot bootseq variable isn't updated if the image set as primary fails to boot. Hence the active image could be the secondary image and if you rely on the bootseq variable you might overwrite the current booted - and only working - secondary image. Am I right?

ecsv replied:
Yes, the problem is that you might overwrite the only working image when the secondary (fallback) image was booted.

Another option would to search for the 0:HLOS index and compare the positions of inactive and 0:HLOS

mkresin replied:
Not really an improvement. Leave it as it is and document somewhere why it has do be done this way (commit message, comment...).

mkresin commented on target/linux/ipq806x/base-files/lib/upgrade/openmesh.sh:
```diff
@@ -0,0 +1,87 @@
+
+# make sure we got uboot-envtools and fw_env.config copied over to the ramfs
+# create /var/lock for the lock "fw_setenv.lock" of fw_setenv
+platform_add_ramfs_ubootenv()
```
This shouldn't be required. `target/linux/ipq806x/base-files/lib/upgrade/platform.sh` has already the required commands to copy `fw_*` to the ramfs:

```
RAMFS_COPY_BIN='fw_printenv fw_setenv'
RAMFS_COPY_DATA='/etc/fw_env.config /var/lock/fw_printenv.lock'
```

ecsv replied:
Looks like it should work but it didn't work for me (stopped because /var/lock/fw_printenv.lock was not available). Can check later what is wrong

mkresin replied:
Yes please check why it doesn't work. If it is a bug, other images within the target are affected as well.

ecsv replied:
This is what I see during an upgrade without the copying

    root@OpenWrt:/# sysupgrade /tmp/openwrt-ipq806x-openmesh_a42-squashfs-sysupgrade.bin
    Saving config files...
    Commencing upgrade. All shell sessions will be closed now.
    Watchdog handover: fd=3
    - watchdog -
    killall: telnetd: no process killed
    Sending TERM to remaining processes ... ntpd ubusd logd netifd odhcpd dnsmasq
    Sending KILL to remaining processes ...
    Switching to ramdisk...
    Performing system upgrade...
    2883584+0 records in
    2883584+0 records out
    Unlocking inactive ...

    Writing from <stdin> to inactive ...
    Appending jffs2 data from /tmp/sysupgrade.tgz to inactive..                                                                                                                                                                                                                     
    Error opening lock file /var/lock/fw_printenv.lock
    failed to update U-Boot environment
    Upgrade completed
    Rebooting system...

The system will then reboot and return to the old flash region (the one which wasn't updated). It looks to me like the binary was copied but for some reason the /var/lock/fw_printenv.lock is not working. And I also found that linksys.sh is manually creating this file:

	mkdir -p /var/lock
	local part_label="$(linksys_get_target_firmware)"
	touch /var/lock/fw_printenv.lock

So I am not the only one experiencing this kind of problem. My assumption is now that:

1. fw_printenv.lock doesn't exist yet in /var/lock
2. sysupgrade is started
3. install_file doesn't find the source file
4. install_file skips the creation of /var/log

Just check the code of install_file

    install_file() { # <file> [ <file> ... ]
    	local target dest dir
    	for file in "$@"; do
    		if [ -L "$file" ]; then
    			target="$(readlink -f "$file")"
    			dest="$RAM_ROOT/$file"
    			[ ! -f "$dest" ] && {
    				dir="$(dirname "$dest")"
    				mkdir -p "$dir"
    				ln -s "$target" "$dest"
    			}
    			file="$target"
    		fi
    		dest="$RAM_ROOT/$file"
    		[ -f "$file" -a ! -f "$dest" ] && {
    			dir="$(dirname "$dest")"
    			mkdir -p "$dir"
    			cp "$file" "$dest"
    		}
    	done
    }

The last check `[ -f "$file" -a ! -f "$dest" ]` will not be started when the src file doesn't exist. And this is the reason why the `mkdir -p` in this block is not started.

I've tested this now by changing the function to only create the directory:

    # create /var/lock for the lock "fw_setenv.lock" of fw_setenv
    # the rest is copied using ipq806x's RAMFS_COPY_BIN and RAMFS_COPY_DATA
    platform_add_ramfs_ubootenv()
    {
            mkdir -p $RAM_ROOT/var/lock
    }
    append sysupgrade_pre_upgrade platform_add_ramfs_ubootenv
bootenv

This version now work perfectly fine (and I've added this version for now to my PR).

@NeoRaider you've worked on that code a while back. Maybe you have an opinion how this should work correctly.

Right now, the users of fw_setenv (allnet openmesh and linksys) seem to have a private `mkdir -p $RAMFS/var/lock` in their scripts. Either using a variant of `platform_add_ramfs_ubootenv` or a simple `mkdir -p` somewhere before they call fw_setenv

NeoRaider replied:
We could simply add the `mkdir -p $RAM_ROOT/var/lock` to switch_to_ramfs() in `/lib/upgrade/stage2` rather than duplicating it in every target that needs it.

jow- replied:
I would love if we could somehow share the creation of these essential volatile directories with the mkdir commands in `/etc/init.d/boot` - maybe fold the mkdir commands there into a `/lib/function.sh` procedure we can call both from init and sysupgrade stage2?

mkresin commented on target/linux/ipq806x/image/Makefile:
```diff
@@ -294,4 +294,21 @@ define Device/AP-DK04.1-C1
+	DEVICE_TITLE := OpenMesh A42
+	KERNEL = kernel-bin | lzma | fit lzma $$(DTS_DIR)/$$(DEVICE_DTS).dtb | pad-to 64k
+	IMAGE_SIZE := 15616k
+	IMAGES = factory.bin sysupgrade.tar
```
please use sysupgrade.bin as image extension.

mkresin commented on target/linux/ipq806x/image/Makefile:
```diff
@@ -294,4 +294,21 @@ define Device/AP-DK04.1-C1
+	BLOCKSIZE := 64k
+	SUPPORTED_DEVICES := openmesh,a42
+	DEVICE_TITLE := OpenMesh A42
+	KERNEL = kernel-bin | lzma | fit lzma $$(DTS_DIR)/$$(DEVICE_DTS).dtb | pad-to 64k
```
Please use `pad-to $$$$(BLOCKSIZE)`

And I finally get why it is required.  the rootfs is padded on it's own and doesn't fix an odd padding of the kernel, as it is done with the usual kernel+rootfs images.

ecsv commented:
I went to the list and did everything except the RAMFS changes. I will have a look at that - so please **don't merge it yet**.

The changes are now:

* added documentation why there is a hardcoded mtdindex in the sysupgrade code
* remove the manual shell invocation in ce-image helper code
* don't ignore return codes in ce-image helper code
* renamed Build/combined-ext-image to Build/openmesh-image
* modified kernel padding to use $${BLOCKSIZE} instead of hardcoding 64k
* renamed sysupgrade.tar to sysupgrade.bin

ecsv commented:
The sysupgrade_pre_upgrade was changed to only create the /var/lock directory. See https://github.com/openwrt/openwrt/pull/633#discussion_r160421147 for more information.

mkresin commented:
Merged into the ipq806x branch of [my staging tree](https://git.openwrt.org/?p=openwrt/staging/mkresin.git;a=summary).

I had to rebase your PR on top of another. Please have a look/test to ensure everything is still as it should be.

The `$RAM_ROOT/var/lock` issue is out of scope and should be fixed independently from this PR. Lets see if someone picks up the task.

ecsv commented:
Not sure when I have time to really test it. Here are the changes from the rebase (just for documentation purposes)

* ipq-wifi conflict was resolved (rt-ac58u and nbg6617 were previously dropped from the package)
* lib/ipq806x.sh changes were removed (file doesn't exist anymore)
* conflict in 0069-arm-boot-add-dts-files.patch was resolved
* conflicts with the patch "ipq806x: Sort occurrences of boardame alphabetically" were resolved by adding the new block to the correct (as in alphabetically) place

I did the rebase myself and came to the same result. The only possible problem I see is that the board_name function doesn't return the correct string (very unlikely). But I can test that later.

ecsv commented:
I did some basic tests with the rebased version and I didn't find any regressions.

ecsv commented:
The A42 BDFs are now also in the QCA4019 board-2.bin https://github.com/kvalo/ath10k-firmware/commit/5cd2bacd8e22217335e539b416d0cb8d59f8e312

mkresin commented:
@ecsv Thanks a lot for the update!

Does it mean we can update ath10k-firmware to a more recent hash and drop the A42 from ipq-wifi?

ecsv commented:
At least in theory. We just have to make sure that /lib/firmware/ath10k/QCA4019/hw1.0/board-2.bin is really in the image.

mkresin commented:
>At least in theory. We just have to make sure that /lib/firmware/ath10k/QCA4019/hw1.0/board-2.bin is really in the image.

According to my reading should work: https://git.openwrt.org/?p=openwrt/openwrt.git;a=blob;f=package/firmware/ath10k-firmware/Makefile;h=94e59537834a67e8b9d469f6a050697e6cf24789;hb=HEAD#l220

Would you mind to give a try?

ecsv commented:
Yes, PR was now added as #706. I just wanted to test it with an actual A42 (also to the requested tests from Kalle Valo)

ecsv commented:
I was just asked how to boot the device with the ramdisk image. Here is how I tend to do it (fdt_high is rather important here):

    setenv serverip 192.168.2.127
    setenv ipaddr 192.168.2.3
    setenv bootargs 'loglevel=8 earlycon=msm_serial_dm,0x78af000 console=ttyMSM0,115200 mtdparts=spi0.0:256k(0:SBL1),128k(0:MIBIB),384k(0:QSEE),64k(0:CDT),64k(0:DDRPARAMS),64k(0:APPSBLENV),512k(0:APPSBL),64k(0:ART),15616k(inactive),15616k(inactive2)'
    tftpboot 0x84000000 openwrt-ipq806x-openmesh_a42-initramfs-fit-uImage.itb
    set fdt_high 0x85000000
    bootm 0x84000000

ecsv commented:
@mkresin Just saw that you started to move the code for A42 to a new target ipq40xx (https://git.openwrt.org/?p=openwrt/staging/mkresin.git;a=shortlog;h=refs/heads/ipq40xx). Please also make sure that you move CONFIG_GPIO_WATCHDOG from ipq806x to ipq40xx. See https://github.com/openwrt/openwrt/commit/c242643d1aba0f065b4a465cd78f0c54dd7cb3b2 for the commit which introduced it in ipq806x for A42.

mkresin commented:
@ecsv the upcoming ipq40xx target isn't my work. John has the lead and I only nag him with issues I found in the ipq806x/ipq40xx split. Nevertheless, I added CONFIG_GPIO_WATCHDOG to my nag list.

ecsv commented:
I was just asked in private about following situation:

* device checked in cloudtrax from China (and downloaded new configuration)
* SSH is accessible (User: root + Password configured in cloudtrax, IP taken from cloudtrax AP overview page)
* u-boot is not accessible (just reboots when trying to open the u-boot shell)
* u-boot asks for fwupgrade.cfg.sig when trying to upload factory image using ap51-flash

This means that the device is unlocked but still has the signature check key for installed. I don't know whether this was intended by Datto/OpenMesh or not. But you just have to disable the key in this case to make u-boot accessible again:

    if grep KEYS /proc/mtd > /dev/null; then
      dd if=/dev/zero of=/dev/mtdblock9 conv=notrunc bs=1 seek=0 count=8
    else
      dd if=/dev/zero of=/dev/mtdblock7 conv=notrunc bs=1 seek=32768 count=8
    fi

A new OpenWrt image should be installed immediately to avoid that firmware-ng reinstalls the key again. I don't actually know whether it would do that in this situation but the disabling of the key works (see Datto's u-boot GPL source code for more details).

pepe2k commented:
@ecsv here is also related OpenWrt forum topic: https://forum.openwrt.org/t/openwrt-installation-on-openmesh-a62-router/23986

Cheers,
Piotr
